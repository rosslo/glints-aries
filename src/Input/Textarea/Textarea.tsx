import * as React from 'react';

import classNames from 'classnames';
import isFunction from 'lodash/isFunction';

import {
  TextareaContainer,
  TextareaInput,
  TextareaLabel,
} from './TextareaStyle';

const MIN_ROWS = 4;
const MAX_ROWS = 12;

const Textarea: React.FunctionComponent<Props> = props => {
  const {
    label,
    value,
    status,
    disabled,
    className,
    onBlur,
    onChange,
    removeFloatingLabel,
    small,
    ...restProps
  } = props;

  const [floating, setFloating] = React.useState<boolean>(false);
  const [rows, setRows] = React.useState<number>(MIN_ROWS);
  const innerRef = React.useRef(null);
  const textareaInputRef = props.forwardedRef || innerRef;

  const [textareaMaxHeight, setTextareaMaxHeight] = React.useState<number>(0);

  React.useLayoutEffect(() => {
    const textarea = textareaInputRef.current;

    if (!textarea) return;

    setTextareaMaxHeight(~~(textarea.offsetHeight * 3.7 + 23 * 8 + 1));
  }, [textareaInputRef]);

  const handleChange = React.useCallback(
    (e: React.ChangeEvent<HTMLTextAreaElement>) => {
      const previousRows = e.target.rows;
      e.target.rows = MIN_ROWS;

      const currentRows = ~~(e.target.scrollHeight / 30);

      if (currentRows === previousRows) {
        e.target.rows = currentRows;
      }

      if (currentRows >= MAX_ROWS) {
        e.target.rows = MAX_ROWS;
        e.target.scrollTop = e.target.scrollHeight;
      }

      setRows(currentRows < MAX_ROWS ? currentRows : MAX_ROWS);

      if (isFunction(onChange)) {
        return onChange(e);
      }
    },
    [onChange]
  );

  const handleFocusChange = React.useCallback(
    (e: React.FocusEvent<HTMLTextAreaElement>) => {
      setFloating(e.target.value.length > 0);

      if (onBlur !== undefined) {
        return onBlur(e);
      }
    },
    [onBlur]
  );

  React.useEffect(() => {
    if (value && value !== '' && !floating) {
      setFloating(true);
    }
  }, [value, floating]);

  return (
    <TextareaContainer className={classNames('aries-textarea', className)}>
      <TextareaInput
        ref={textareaInputRef}
        placeholder={removeFloatingLabel && label}
        rows={rows}
        status={status}
        disabled={disabled}
        onBlur={handleFocusChange}
        onChange={handleChange}
        floating={floating}
        value={value}
        aria-label={label}
        small={small}
        {...restProps}
        style={{
          maxHeight: `${textareaMaxHeight}px`,
        }}
      />
      {!removeFloatingLabel && (
        <TextareaLabel
          data-testid="textarea-label"
          floating={floating}
          status={status}
          small={small}
        >
          {label}
        </TextareaLabel>
      )}
    </TextareaContainer>
  );
};

interface Props extends React.ComponentPropsWithoutRef<typeof TextareaInput> {
  label?: string;
  onBlur?(e: React.FocusEvent<HTMLTextAreaElement>): void;
  onChange?(e: React.ChangeEvent<HTMLTextAreaElement>): void;
  removeFloatingLabel?: boolean;
  forwardedRef?: React.RefObject<HTMLTextAreaElement>;
}

const forwardRef = (
  props: Props,
  ref: React.RefObject<HTMLTextAreaElement>
) => <Textarea {...props} forwardedRef={ref} />;

forwardRef.displayName = Textarea.name;

export default React.forwardRef(forwardRef);
