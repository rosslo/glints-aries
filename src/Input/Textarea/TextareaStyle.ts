import styled from 'styled-components';
import { SecondaryColor, Greyscale } from '../../Utils/Colors';

export const TextareaContainer = styled.div`
  flex-grow: 1;
  position: relative;
  display: flex;
`;

export const TextareaLabel = styled.label<TextareaLabelProps>`
  position: absolute;
  left: 22px;
  top: 1em;
  color: #666666;
  transition: all 0.2s;
  pointer-events: none;
  font-weight: 400;
  font-size: 16px;

  ${({ floating }) => {
    if (floating) {
      return `
        padding: 0 5px;
        transform: translate3d(-15px,-20px,0);
        transition: all .2s;
        font-size: 12px;
        background: ${Greyscale.white}
      `;
    }
  }}
`;

interface TextareaLabelProps {
  floating: boolean;
  status?: string;
  small?: boolean;
}

export const TextareaInput = styled.textarea<TextareaInputProps>`
  position: relative;
  width: 100%;
  outline: none;
  font-size: 16px;
  line-height: 1.5;
  padding: ${({ small }) => (small ? '8px 16px' : '12px 16px')};
  background-color: ${({ status }) =>
    status === 'error' ? ' rgba(236, 39, 43, 0.08)' : 'rgb(1, 126, 183, 0.08)'};
  border: unset;
  transition: border 0.5s;
  resize: vertical;
  overflow: auto;
  height: auto;
  border-radius: ${({ small }) => (small ? '4px' : '8px')};

  &:disabled {
    cursor: not-allowed;
    background: ${SecondaryColor.whitesmoke};

    + ${TextareaLabel} {
      background: transparent;
      color: ${SecondaryColor.grey};
    }

    &:hover {
      border: 2px solid ${SecondaryColor.lightgrey};
    }
  }

  &:hover {
    background-color: ${({ status }) =>
      status === 'error'
        ? ' rgba(236, 39, 43, 0.08)'
        : 'rgb(1, 126, 183, 0.16)'};
    border: unset;

    + ${TextareaLabel} {
      color: #666666;
    }
  }

  &:focus {
    background-color: ${({ status }) =>
      status === 'error'
        ? ' rgba(236, 39, 43, 0.08)'
        : 'rgb(1, 126, 183, 0.24)'};
    border: unset;

    + ${TextareaLabel} {
      padding: 0 5px;
      transform: translate3d(-15px, -20px, 0);
      transition: all 0.2s;
      color: #666666;
      font-size: 12px;
      background: ${Greyscale.white};
    }
  }
`;

interface TextareaInputProps {
  floating?: boolean;
  status?: string;
  small?: boolean;
}
