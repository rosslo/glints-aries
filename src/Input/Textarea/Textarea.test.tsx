import * as React from 'react';

import * as renderer from 'react-test-renderer';
import '@testing-library/jest-dom/extend-expect';
import { render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import Textarea from './Textarea';
import { SecondaryColor, PrimaryColor } from '../../Utils/Colors';

interface Props {
  label: string;
  value: string;
  onChange: jest.Mock;
}

const props: Props = {
  label: 'Description',
  value: 'Explanation',
  onChange: jest.fn().mockImplementation(event => event.target.value),
};

const inputValue =
  'As the first sign in the zodiac, the presence of Aries always marks the beginning of something energetic and turbulent.';

function setupTextarea(props: Props, otherProps?: any) {
  const { getByLabelText, ...utils } = render(
    <Textarea label={props.label} onChange={props.onChange} {...otherProps} />
  );
  const textareaInput = getByLabelText(props.label) as HTMLTextAreaElement;
  return { textareaInput, utils };
}

function setupTextareaWithChangeEvent(props: Props) {
  const { textareaInput } = setupTextarea(props);
  userEvent.type(textareaInput, inputValue);
  return { textareaInput };
}

function createNodeMock(element: any) {
  if (element.type === 'textarea') {
    return {
      value: 0,
      offsetHeight: 34,
    };
  }
  return null;
}

it('<Textarea> should render with a value and label of Description', () => {
  const options = { createNodeMock };
  const TextareaSnapshot = renderer
    .create(<Textarea label={props.label} value={props.value} />, options)
    .toJSON();
  expect(TextareaSnapshot).toMatchSnapshot();
});

describe('when it is rendered', () => {
  it('should display the correct input label', () => {
    const { textareaInput } = setupTextarea(props);
    expect(textareaInput).toBeTruthy();
  });

  it('should have an empty string as its value', () => {
    const { textareaInput } = setupTextarea(props);
    expect(textareaInput.value).toEqual('');
  });
});

describe('when a value is entered', () => {
  it('should call onChange for each input', () => {
    setupTextareaWithChangeEvent(props);
    expect(props.onChange).toHaveBeenCalledTimes(inputValue.length);
  });

  it('should return the input value when onChange is passed a function: event => event.target.value', () => {
    setupTextareaWithChangeEvent(props);
    expect(props.onChange.mock.results.slice(-1)[0].value).toEqual(inputValue);
  });

  it('should display the correct value', () => {
    const { textareaInput } = setupTextareaWithChangeEvent(props);
    expect(textareaInput.value).toEqual(inputValue);
  });
});

describe('when status is:', () => {
  it("'error', it should display a red background", () => {
    const { textareaInput } = setupTextarea(props, { status: 'error' });
    expect(textareaInput).toHaveStyle(`
      background-color: rgba(236, 39, 43, 0.08);
    `);
  });
});

describe('when removeFloatingLabel is true', () => {
  it('should remove the floating label', () => {
    const { textareaInput } = setupTextarea(props, {
      removeFloatingLabel: true,
    });
    const textFieldLabel = textareaInput.parentElement.querySelector(
      '[data-testid="textarea-label"]'
    ) as HTMLLabelElement;
    expect(textareaInput.parentElement).not.toContainElement(textFieldLabel);
  });

  it('should display a placeholder', () => {
    const { utils } = setupTextarea(props, { removeFloatingLabel: true });
    const textareaInput = utils.getByPlaceholderText(props.label);
    expect(textareaInput).toBeTruthy();
  });
});

describe('<Textarea /> forwards ref to underlying textarea element', () => {
  const label = 'forward-ref';

  test('ref is being forwarded correctly', () => {
    const ref = React.createRef<HTMLTextAreaElement>();
    const { queryByLabelText } = render(<Textarea ref={ref} label={label} />);
    const textarea = queryByLabelText(label);
    expect(textarea).toEqual(ref.current);
  });

  test('when value is passed, textareaLabel style will follow when floating is true', () => {
    const { getByTestId } = render(
      <Textarea label={label} value={props.value} />
    );
    const textareaLabel = getByTestId('textarea-label');

    expect(textareaLabel).toHaveStyle(`padding: 0 5px;
    transform: translate3d(-15px,-20px,0);
    transition: all .2s;
    font-size: 12px;`);
  });

  test('underlying textarea can be focused/blurred through ref', () => {
    const ref = React.createRef<HTMLTextAreaElement>();
    const onBlur = jest.fn();
    const { queryByLabelText } = render(
      <Textarea ref={ref} label={label} onBlur={onBlur} />
    );
    const textarea = queryByLabelText(label);
    expect(textarea).not.toHaveFocus();

    ref.current.focus();
    expect(textarea).toHaveFocus();

    ref.current.blur();
    expect(textarea).not.toHaveFocus();
    expect(onBlur).toHaveBeenCalledTimes(1);
  });
});
