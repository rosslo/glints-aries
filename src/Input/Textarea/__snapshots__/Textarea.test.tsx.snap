// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Textarea> should render with a value and label of Description 1`] = `
<div
  className="TextareaStyle__TextareaContainer-sc-1998n74-0 RSaAJ aries-textarea"
>
  <textarea
    aria-label="Description"
    className="TextareaStyle__TextareaInput-sc-1998n74-2 hNgSpA"
    onBlur={[Function]}
    onChange={[Function]}
    rows={4}
    style={
      Object {
        "maxHeight": "310px",
      }
    }
    value="Explanation"
  />
  <label
    className="TextareaStyle__TextareaLabel-sc-1998n74-1 kLOSMy"
    data-testid="textarea-label"
  >
    Description
  </label>
</div>
`;
